pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/ExtensibleLoadManagerImpl.java

```diff
@@ -0,0 +1,225 @@
+        this.brokerRegistry.start();
+        this.serviceUnitStateChannel.start();
+
+        // TODO: Start the load data store.
+
+        this.context = LoadManagerContextImpl.builder()
+                .configuration(conf)
+                .brokerRegistry(brokerRegistry)
+                .brokerLoadDataStore(null)
+                .topBundleLoadDataStore(null).build();
+        // TODO: Start load data reporter.
+
+        // TODO: Start unload scheduler and bundle split scheduler
+        this.started = true;
+    }
+
+    @Override
+    public synchronized void initialize(PulsarService pulsar) {
+        this.pulsar = pulsar;
+        this.conf = pulsar.getConfiguration();
+    }
+
+    @Override
+    public CompletableFuture<Optional<BrokerLookupData>> assign(Optional<ServiceUnitId> topic,
+                                                                ServiceUnitId serviceUnit) {
+
+        final String bundle = serviceUnit.toString();
+
+        CompletableFuture<Optional<BrokerLookupData>> future = lookupRequests.computeIfAbsent(bundle, k -> {
+            final CompletableFuture<Optional<String>> owner;
+            // Assign the bundle to channel owner if is internal topic, to avoid circular references.
+            if (topic.isPresent() && isInternalTopic(topic.get().toString())) {
+                owner = serviceUnitStateChannel.getChannelOwnerAsync();
+            } else {
+                owner = serviceUnitStateChannel.getOwnerAsync(bundle).thenCompose(broker -> {
+                    // If the bundle not assign yet, select and publish assign event to channel.
+                    if (broker.isEmpty()) {
+                        return this.selectAsync(serviceUnit).thenCompose(brokerOpt -> {
+                            if (brokerOpt.isPresent()) {
+                                log.info("Selected new owner broker: {} for bundle: {}.", brokerOpt.get(), bundle);
+                                return serviceUnitStateChannel.publishAssignEventAsync(bundle, brokerOpt.get())
+                                        .thenApply(Optional::of);
+                            } else {
+                                throw new IllegalStateException(
+                                        "Failed to select the new owner broker for bundle: " + bundle);
+                            }
+                        });
+                    }
+                    // Already assigned, return it.
+                    return CompletableFuture.completedFuture(broker);
+                });
+            }
+
+            return owner.thenCompose(broker -> {
+                if (broker.isEmpty()) {
+                    String errorMsg = String.format(
+                            "Failed to look up a broker registry:%s for bundle:%s", broker, bundle);
+                    log.error(errorMsg);
+                    throw new IllegalStateException(errorMsg);
+                }
```

Comment:
Throwing an unchecked exception in the callback of `CompletableFuture` is not a good behavior. The caller usually prints the stack of the exception (in `exceptionally` or `whenComplete`). But the stack is meaningless. A better practice is to print an error log here and return an empty `Optional`.

Reply:
For example, in this method, the caller side only needs to handle if the assign operation succeeded. If you throw an exception here, the caller side should process it like:

```java
if (e instanceof IllegalStateException) {
    log.error("Failed to assign: {}", e.getMessage());
} else {
    log.error("Unexpected exception in assign", e);
}
```

Reply:
The caller handled this exception, see: https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java#L315-L317

And in the old logic, it also throws the `IllegalStateException`, https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java#L398-L400

Reply:
The exception is handled for `NamespaceService#getBrokerServiceUrlAsync`, are they equivalent?

BTW, you also throws an `IllegalStateException` in line 136, which means `broker.isEmpty()` is always false here. Even if `ExtensibleLoadManager#assign` is equivalent and will replace `NamespaceService#getBrokerServiceUrlAsync`, the line 136 should return an empty `Optional` to go to the if branch of line 147.